Finally added #1003

Notarin opened the pull request:
Wow this was sitting on my agenda for AWHILE.

Notarin commented:
First off, I noticed an issue, one that I may not be able to easily resolve myself. I need to convert this into a draft because I myself now think this is no longer ready for rebase/merge.
Second, Thank you for the input. I will try, but it already took me a year to do this (crippling executive dysfunction is a bitch), honestly I am surprised no one did this before me.

- I will make this change and force push
- where? the commit message? the PR title? the PR description? in a PR comment?
- I will try, but I am not at all confident I will be able to. I based my PR on a previous PR of a nearly identical keyboard.

I noticed the volume wheel and the mute button are not interfacing properly. Is there docs or instructions on identifying these keys and how to handle them?
Is there docs on the steps to reflect these changes within the GUI?
I am also requesting that someone with a `K70 RGB TKL`(non-OPX) test my PR to ensure I have not committed any regressions. @minneyar, would you be willing? Thank you.

I am going to convert this into a draft until I, or someone else, fixes the volume keys issue. Thank you.

gtjoseph commented:
> * where? the commit message? the PR title? the PR description? in a PR comment?

Ideally the commit message and the PR title/description should match.  When you first submit a PR on GitHub, the first line of the commit message becomes the title and the rest of the commit message becomes the PR description.  However, when you force push your branch to update the PR, the PR title and description don't auto-update so you have to do that manually.  So add the Resolves line to your commit message then after you've forced push, change the PR title and description to match.

> * I will try, but I am not at all confident I will be able to. I based my PR on a previous PR of a nearly identical keyboard.

Fair enough.  Someone can create a follow-on change to update the GUI.  Possibly me but I also suffer from the same executive dysfunction :)

> I noticed the volume wheel and the mute button are not interfacing properly.
> I am also requesting that someone with a `K70 RGB TKL`(non-OPX) test my PR to ensure I have not committed any regressions.

They volume wheel and mute button are working fine for me with your PR.  My K70 KTL Champion Cherry MX also works fine with your PR.  It should since you didn't change anything with the existing K70 TKL support.

> I am going to convert this into a draft until I, or someone else, fixes the volume keys issue. Thank you.

Just FYI...  This morning, before I saw your PR, I created a patch to add the keyboard.  The only difference between your and mine was that yours used P_K70_TKL_OPX and mine used P_K70_TKL_OPT. :)

gtjoseph commented:
Oh for the wheel and mute button...  What keyboard were you using before?  What Linux distribution and desktop environment are you using?  

The volume wheel sends the standard XF86AudioRaiseVolume and XF86AudioLowerVolume events and the mute button sends the standard XF86AudioMute event so they should work without any customization.

Notarin commented:
> Ideally the commit message and the PR title/description should match.

> first line of the commit message becomes the title and the rest of the commit message becomes the PR description.

Thank you, I will force-push a new title, and I shall update the PR title and description accordingly. Obviously I would need to do so manually, it would be rather silly if it automatically updated the PR title/description itself, major UX no-no.

> Someone can create a follow-on change to update the GUI. Possibly me

I will see if I can do it myself, who knows, I might be able to. But if you beat me to it, I see no foul.

> They volume wheel and mute button are working fine for me with your PR.

Have you tested with an OPX keyboard? I may have had a personal skill issue last I used this, but there was an issue regardless. I will hopefully test again and we shall see for sure.

> before I saw your PR, I created a patch to add the keyboard

That's very silly, but honestly I expected it to happen haha.

> Oh for the wheel and mute button... What keyboard were you using before? What Linux distribution and desktop environment are you using?

Well, at the time I was of course using the keyboard relevant to the PR, the "K70 RGB TKL CHAMPION SERIES Optical". I still use the same keyboard, however I have since switched operating systems and subsequently have yet to set up ckb-next on my new system. At the time I was using the OS, Arch. Then, and now, I use hyprland.

> The volume wheel sends the standard XF86AudioRaiseVolume and XF86AudioLowerVolume events and the mute button sends the standard XF86AudioMute event so they should work without any customization

Apologies for my lack of firm understanding, but the issue was not the events not being interpreted correctly, but rather the driver was not recognizing the keys *at all*. While the keys did show up in the GUI, changes made had no effects, furthermore, on the physical keyboard, the LEDs did not light up, and pressing them sent absolutely nothing, as confirmed by `wev` and `xev`, along with some other keyboard diagnostic tool I used. Turning off the ckb-next daemon caused the keys to immediately function as expected. Turning it on, of course resumed the cessation of functionality regarding the keys.

Thank you for your regression test. C++ is not my native tongue, and I had modified I think if I recall a switch-case block. I wanted to ensure I did not misunderstand what it does and nuke its logic. This is why I requested a regression test, because I suspected I may have possibly affected the logic of another keyboard. I figured I did not, aamof I was pretty damn sure I did not, but it would be disrespectful of me to be aware of the possibility I have introduced a regression without so much as mentioning it.
Oh and, ah, found it, see line 37: https://github.com/ckb-next/ckb-next/blob/545c3993c4e82590d15ff9de52a2820eb6f354c0/src/daemon/usb_bragi.c#L25-L42
I am pretty confident that order does not matter in a C++ switch-case block, but I do not know for *sure*, and I *did* usurp the last line position. So I requested a regression test.

gtjoseph commented:
> Have you tested with an OPX keyboard? I may have had a personal skill issue last I used this, but there was an issue regardless. I will hopefully test again and we shall see for sure.

I'm not sure what the OPX designation means specifically but I have the following two keyboards...
CORSAIR K70 RGB TKL CHAMPION SERIES Optical Mechanical Gaming Keyboard ProductId: 1bb9
CORSAIR K70 RGB TKL CHAMPION SERIES Mechanical Gaming Keyboard ProductId: 1b73
Both work fine with your PR.  FYI though, I don't actually use the ckb gui for anything.  I don't even start it.  I'm on Fedora 41 XFCE/X11.

> I am pretty confident that order does not matter in a C++ switch-case block, but I do not know for sure, and I did usurp the last line position. So I requested a regression test.

No worries.  In this case, everything from P_K57_U to P_K70_TKL_OPX falls through to the code that sets `kb->bragi_out_ep = 0x1;` and `kb->bragi_in_ep = 0x82;` so yeah, in this case, order doesn't matter.

gtjoseph commented:
I've submitted PR #1157 to replace this PR so this one can be closed.

tatokis commented:
Thank you for taking the time to work on this
